src/extra/components/xgm.py

```diff
@@ -0,0 +1,526 @@
+    DEVICE_CLASSES = ["DoocsXGM", "DoocsXGMD", "DoocsXGMReduced"]
+
+    def __init__(self, run, device=None, default_sase=None):
+        """
+        Args:
+            run (extra_data.DataCollection): A run containing at least one XGM.
+            device (str, SourceData, KeyData): Specify an XGM to use, necessary
+                if a run contains more than one XGM. This can be any of:
+
+                  - The device name of the control source.
+                  - A `SourceData` or [KeyData][extra_data.KeyData] of either
+                    the control source (e.g. `SA2_XTD1_XGM/XGM/DOOCS`) or
+                    instrument source (e.g. `SA2_XTD1_XGM/XGM/DOOCS:output`) of
+                    an XGM.
+                  - The alias name of either a `SourceData` or
+                    [KeyData][extra_data.KeyData] belonging to an XGM.
+                  - A unique (case-insensitive) substring of an XGM source
+                    name. For example if a run contains `HED_XTD6_XGM/XGM/DOOCS`
+                    and `SA2_XTD1_XGM/XGM/DOOCS`, then passing `device="hed"` or
+                    `device="sa2"` will select those XGMs, respectively (passing
+                    `xtd1` or `xtd6` would also work).
+
+            default_sase (int): Only for XGMs with data for multiple SASE's, to
+                specify a default SASE to retrieve data for. This is useful when
+                working with XGMs on SASE 1 or SASE 3 and you're only interested
+                in data for a specific SASE. This setting applies to all the
+                SASE-specific methods of the class, such as
+                [XGM.intensity()][extra.components.XGM.intensity] and
+                [XGM.npulses()][extra.components.XGM.npulses].
+
+                If it is not set the class will try to guess the SASE of the
+                instrument the data was recorded at, and it will throw an error
+                if:
+
+                - It couldn't guess the SASE of the instrument at all.
+                - It did guess the SASE of the instrument, but it's a different
+                  SASE from the XGM selected. For example, if a run recorded at
+                  SQS has both a SASE 1 and SQS XGM and the SQS XGM is selected,
+                  `default_sase` will automatically be set to `3`. But if the
+                  SASE 1 XGM is selected it will throw an error and ask the user
+                  to set `default_sase` explicitly.
+
+                Valid values are: 0, 1, and 3.
+                Passing 0 always selects the main properties of a device,
+                e.g. `data.intensityTD` instead of the SASE-specific ones.
+        """
+        device = _find_xgm(run, device)
+
+        self._control_source = run[device]
+        self._instrument_source = run[f"{device}:output"]
+
+        self._extra_sases = [sase for sase in [1, 3]
+                             if f"pulseEnergy.numberOfSa{sase}BunchesActual" in self._control_source]
```

Comment:
Was just reading old data, it looks like the keys have changed as some point. Old proposal (e.g. p700000/r1) do not have `pulseEnergy.numberOfBunchesActual` or `pulseEnergy.numberOfSa[13]BunchesActual` but only `pulseEnergy.nummberOfBrunches` (yes, with 2 `m` in nummber)

Reply:
and there's a version with proper spelling `pulseEnergy.numberOfBunches` (e.g. p002429)

Reply:
Nice, thanks for catching that :rofl: :sandwich: I'll probably get around to fixing it next week.

Reply:
This reminds me of #23.

Reply:
Fixed in c378ea5.